WalletWasabi/WabiSabi/Client/AmountDecomposer.cs

```diff
@@ -65,12 +65,49 @@ public IEnumerable<Money> Decompose(IEnumerable<Coin> myInputCoins, IEnumerable<
 				}
 			}
 
-			if (remaining >= MinimumAmountPlusFee)
+			if (remaining >= MinimumAmountPlusFee && remainingVsize >= OutputSize)
 			{
-				outputAmounts.Add(remaining - OutputFee);
+				naiveSet.Add(remaining - OutputFee);
+			}
+
+			setCandidates.Add(naiveSet.Count, naiveSet);
+			var before = DateTimeOffset.UtcNow;
+			do
+			{
+				remaining = inputs.Sum();
+				remainingVsize = AvailableVsize;
+				var currSet = new List<Money>();
+				do
+				{
+					var denomPlusFees = denoms.Where(x => x <= remaining && x >= (remaining / 3)).ToList();
+					var denomPlusFee = denomPlusFees.RandomElement();
+					if (denomPlusFee is null || remaining < MinimumAmountPlusFee || remainingVsize < 2 * OutputSize)
+					{
+						break;
+					}
+
+					if (denomPlusFee <= remaining)
+					{
+						currSet.Add(denomPlusFee - OutputFee);
+						remaining -= denomPlusFee;
+						remainingVsize -= OutputSize;
+					}
+				}
+				while (currSet.Count <= naiveSet.Count || currSet.Count <= 3);
+
+				if (currSet.Count <= naiveSet.Count || currSet.Count <= 3)
+				{
+					if (remaining >= MinimumAmountPlusFee && remainingVsize >= OutputSize)
+					{
+						currSet.Add(remaining);
+					}
+
+					setCandidates.TryAdd(currSet.Count, currSet);
+				}
 			}
+			while ((DateTimeOffset.UtcNow - before).TotalMilliseconds <= 30);
 
-			return outputAmounts;
+			return setCandidates.RandomElement().Value;
```

Comment:
I was thinking if we should bias smaller number of outputs. Ultimately, I'm not sure it makes sense as eventually we want to bias this for a different reason: we'll probably want to have a desired number of UTXOs and randomly bias this function later on to either prefer creating more UTXOs or less.

Reply:
https://github.com/zkSNACKs/WalletWasabi/pull/6908/commits/fcd5212c8c8118b3eb158f22a3998be3106f18e5